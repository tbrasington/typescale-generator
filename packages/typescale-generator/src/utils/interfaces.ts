
export interface ViewPortProps {
  width: number;
  fontSize: number;
  typeScale: number;
}

export type TypeScaleProps = {[key: string]: GeneratedItemProps}
export interface GeneratedNamedScalesProps {
  typeScale: TypeScaleProps
}


export interface GeneratedItemProps {
  min: ViewPortProps;
  max?: ViewPortProps;
  clamp: string;
}

export interface RangeProps {
  min: number;
  max: number;
}
export interface GenericKeyValueProps {
  [key: string]: string;
}

export const NamedScales = {
  MAJOR_SECOND: "MAJOR_SECOND",
  MINOR_SECOND: "MINOR_SECOND",
  MAJOR_THIRD: "MAJOR_THIRD",
  MINOR_THIRD: "MINOR_THIRD",
  PERFECT_FOURTH: "PERFECT_FOURTH",
  AUGMENTED_FOURTH: "AUGMENTED_FOURTH",
  PERFECT_FIFTH: "PERFECT_FIFTH",
  GOLDEN_RATIO: "GOLDEN_RATIO",
  OCTAVE: "OCTAVE",
  MAJOR_SIXTH: "MAJOR_SIXTH",
  MINOR_SEVENTH: "MINOR_SEVENTH",
  MAJOR_SEVENTH: "MAJOR_SEVENTH",
};


/*
 * Types for scale enums
 */
export type NamedScalesProps = typeof NamedScales[keyof typeof NamedScales];


export const TypographyScaleValues = {
  [NamedScales.MINOR_SECOND]: { name: "Minor second", value: 1.067 },
  [NamedScales.MAJOR_SECOND]: { name: "Major second", value: 1.125 },
  [NamedScales.MINOR_THIRD]: { name: "Minor third", value: 1.2 },
  [NamedScales.MAJOR_THIRD]: { name: "Major third", value: 1.25 },
  [NamedScales.PERFECT_FOURTH]: { name: "Perfect fourth", value: 1.333 },
  [NamedScales.AUGMENTED_FOURTH]: {
    name: "Augmented fourth",
    value: 1.414,
  },
  [NamedScales.PERFECT_FIFTH]: { name: "Perfect fifth", value: 1.5 },
  [NamedScales.GOLDEN_RATIO]: { name: "Golden ratio", value: 1.618 },
  [NamedScales.MAJOR_SIXTH]: { name: "Major sixth", value: 1.667 },
  [NamedScales.MINOR_SEVENTH]: { name: "Minor seventh", value: 1.778 },
  [NamedScales.MAJOR_SEVENTH]: { name: "Major seventh", value: 1.875 },
  [NamedScales.OCTAVE]: { name: "Octave", value: 2 },
};
