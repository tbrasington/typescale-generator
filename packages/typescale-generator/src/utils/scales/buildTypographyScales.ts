import { ViewPortProps, GeneratedNamedScalesProps, stepValues } from ".";
import { round } from "../round";
import { clampBuilder } from "./clampBuilder";

/**
 * @returns An array of units and clamped values.
 * ```js
 *    {
 *      typeScale = {
 *         "0": {
 *           clamp : 'clamp(1rem, 0.93rem + 0.36vw, 1.25rem)',
 *           min : {
 *              fontSize: 14.05,
 *              width: 320,
 *              scale : 1.25
 *          },
 *          max : {
 *              fontSize: 17.57,
 *              width: 1440,
 *              scale : 1.25
 *          }
 *        }
 *      }
 *    }
 * ```
 * @example 
 * ```js
*   buildTypographyScales({
      min: {
        width: 320,
        fontSize: 16,
        typeScale: TypographyScaleValues[scale].value,
      },
      max: {
        width: 1440,
        fontSize: 20,
        typeScale: TypographyScaleValues[scale].value,
      },
     range: [-2, -1, 0, 1, 2, 3, 4, 5],
  });
  ```
 */
export function buildTypographyScales(opts: {
  range: number[];
  min: ViewPortProps;
  max?: ViewPortProps;
}) {
  const { min, max, range } = opts;

  const system: GeneratedNamedScalesProps = {
    typeScale: [],
  };

  const minFluidTypeStep = range[0];
  const maxFluidTypeStep = range[range.length - 1];
  for (let i = minFluidTypeStep; i <= maxFluidTypeStep; i++) {
   
    system.typeScale[i] = stepValues({
      min: min,
      max: max,
      step: i,
    })
  }

  return {
    ...system,
  };
}