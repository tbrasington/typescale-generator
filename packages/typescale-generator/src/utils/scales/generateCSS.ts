import { GeneratedItemProps, GenericKeyValueProps, TypeScaleProps } from "../interfaces";


export function generateCSS({ scales }: { scales: TypeScaleProps; }) {
  const typeSteps: GenericKeyValueProps = {};

  for (const step of Object.keys(scales)) {
    typeSteps[`--${step}`] = `${scales[`${step}`].clamp};`;
  }

  return `:root ${JSON.stringify(typeSteps, null, "  ").replace(/"/g, "").replace(/,/g, "")}`;
}
