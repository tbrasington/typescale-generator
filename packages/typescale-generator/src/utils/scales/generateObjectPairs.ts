import { GenericKeyValueProps, TypeScaleProps } from "../interfaces";

/**
 * @returns An object of the steps with the min and max unit values, paired with the maxwidth.
 * @example ```js
 *
 * ```
 */

export function generateObjectPairs({ scales }: { scales: TypeScaleProps }) {
   
  const remap = Object.keys(scales).map((item) => {
 
    return {
      step: item,
      fontSize: {
        ...scales[item].sizes
      },
    };
  });

  return remap;
}
