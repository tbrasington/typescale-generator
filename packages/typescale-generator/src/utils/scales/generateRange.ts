import { RangeProps } from ".";

/** 
* 
* Generates a range of numbers to use as the steps in the typography scale
* @returns [-2, -1, 0, 1, 2, 3, 4, 5]
*/
export function generateRange({min, max} : RangeProps)  {
  return Array.from(new Array(max + 1 + -1 * min), (_x, i) => i + min);
}
