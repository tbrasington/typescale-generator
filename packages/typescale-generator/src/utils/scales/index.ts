export { generateRange } from "./generateRange";
export { stepValues } from "./stepValues";
export { numberToRem } from "./numberToRem";
export { clampBuilder } from "./clampBuilder";
export { buildTypographyScales } from "./buildTypographyScales";
export { generateObjectPairs } from "./generateObjectPairs";
export { generateCSS } from './generateCSS'
export { generateObject } from './generateObject'