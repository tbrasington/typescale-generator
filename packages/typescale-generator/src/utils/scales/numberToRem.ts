export function numberToRem(px: number): string {
  return `${px / 16}rem`;
}