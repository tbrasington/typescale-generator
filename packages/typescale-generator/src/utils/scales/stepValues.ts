import { ViewPortProps } from "../interfaces";
import { round } from "../round";
import { clampBuilder } from "./clampBuilder";
import { numberToRem } from "./numberToRem";


/**
 * @returns An object of CSS clamped values and their min and max number values, corresponding to their step.
 * ```js
 *    { 
 *       clamp : 'clamp(1rem, 0.93rem + 0.36vw, 1.25rem)',
 *       min : {
 *        fontSize: 14.05,
 *         width: 320,
 *         scale : 1.25
 *          },
 *          max : {
 *              fontSize: 17.57,
 *              width: 1440,
 *              scale : 1.25
 *          }
 *        }
 *      
 * ```
 * @example 
 * ```js
*   stepValues({
      min: {
        width: 320,
        fontSize: 16,
        typeScale: TypographyScaleValues[scale].value,
      },
      max: {
        width: 1440,
        fontSize: 20,
        typeScale: TypographyScaleValues[scale].value,
      },
      step : 1
  });
  ```
 */
export function stepValues({ step, sizes}: {

  sizes: ViewPortProps[];
  step: number;
}) {

  const min = sizes && sizes[0];
    const valueMin = round(
      min.fontSize * Math.pow(min.typeScale, step)
    );

    // array
    const values = sizes.map((size) => {
      return {...size, fontSize: round(size.fontSize * Math.pow(size.typeScale, step))};
    });


    const maxLength =  sizes.length -1
    const max = sizes && maxLength>=1 && sizes[maxLength]
    const valueMax = max && sizes[maxLength]  ? round(
      sizes[maxLength].fontSize * Math.pow(sizes[maxLength].typeScale, step)
    ) : undefined;

    return {
      step : step,
      sizes: values,
      clamp: valueMax && max ? clampBuilder(
        min.width,
        max.width,
        valueMin,
        valueMax
      ) : numberToRem(valueMin),
    }
}