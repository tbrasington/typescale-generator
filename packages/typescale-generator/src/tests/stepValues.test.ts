import { TypographyScaleValues } from "../utils/interfaces";
import { stepValues } from "../utils/scales/stepValues";

describe("Generate a set of step values", () => {
  test("snapshot with just a min", () => {
    const value = stepValues({
     sizes: [{
      width: 320,
      fontSize: 16,
      typeScale: TypographyScaleValues.MINOR_SECOND.value,
    } ],
      step: 1,
    });
    expect(value).toMatchSnapshot();
  });

  test("min values only", () => {
    const value = stepValues({
     sizes: [{
      width: 320,
      fontSize: 16,
      typeScale: TypographyScaleValues.MINOR_SECOND.value,
    }],
      step: 0,
    });

    expect(value.clamp).toBe("1rem");
    expect(value.sizes[0].fontSize).toBe(16);
    expect(value.sizes[0].width).toBe(320);
    expect(value.sizes[0].typeScale).toBe(1.067);
    expect(value.sizes[1]).toBe(undefined);
  });

  test("min and max values", () => {
    const value = stepValues({
      sizes: [{
      width: 320,
      fontSize: 16,
      typeScale: TypographyScaleValues.MINOR_SECOND.value,
    },{
      width: 1440,
      fontSize: 16,
      typeScale: TypographyScaleValues.MINOR_SECOND.value
    }],
      step: 0,
    });
    expect(value.clamp).toBe('clamp(1rem, 1rem + 0vw, 1rem)');
    expect(value.sizes[0].fontSize).toBe(16);
    expect(value.sizes[0].width).toBe(320);
    expect(value.sizes[0].typeScale).toBe(1.067);
    expect(value.sizes[0]).not.toBe(undefined);
    if (value.sizes[1]) {
      expect(value.sizes[1].fontSize).toBe(16);
      expect(value.sizes[1].width).toBe(1440);
      expect(value.sizes[1].typeScale).toBe(1.067);
    }
  });
});
