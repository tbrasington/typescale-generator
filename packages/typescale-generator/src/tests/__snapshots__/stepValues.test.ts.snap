// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate a set of step values snapshot with just a min 1`] = `
Object {
  "clamp": "1.066875rem",
  "sizes": Array [
    Object {
      "fontSize": 17.07,
      "typeScale": 1.067,
      "width": 320,
    },
  ],
}
`;
